src/ring_theory/polynomial_algebra.lean

```diff
@@ -0,0 +1,342 @@
+
+lemma left_inv (x : A ⊗ polynomial R) :
+  inv_fun R A ((to_fun_alg_hom R A) x) = x :=
+begin
+  apply tensor_product.induction_on _ _ x,
+  { simp [inv_fun], },
+  { intros a p, dsimp only [inv_fun],
+    rw [to_fun_alg_hom_apply_tmul, eval₂_sum],
+    simp_rw [eval₂_monomial, alg_hom.coe_to_ring_hom, algebra.tensor_product.tmul_pow, one_pow,
+      algebra.tensor_product.include_left_apply, algebra.tensor_product.tmul_mul_tmul,
+      mul_one, one_mul, ←algebra.commutes, ←algebra.smul_def'', smul_tmul],
+    rw [finsupp.sum, ←tmul_sum],
+    conv_rhs { rw [←sum_C_mul_X_eq p], },
+    simp only [algebra.smul_def''],
+    refl, },
+  { intros p q hp hq,
+    simp only [alg_hom.map_add, inv_fun_add, hp, hq], },
+end
+
+lemma right_inv (x : polynomial A) :
+  (to_fun_alg_hom R A) (inv_fun R A x) = x :=
+begin
+  apply polynomial.induction_on' x,
+  { intros p q hp hq, simp only [inv_fun_add, alg_hom.map_add, hp, hq], },
+  { intros n a,
+    rw [inv_fun, eval₂_monomial, alg_hom.coe_to_ring_hom, algebra.tensor_product.include_left_apply,
+      algebra.tensor_product.tmul_pow, one_pow, algebra.tensor_product.tmul_mul_tmul,
+      mul_one, one_mul, to_fun_alg_hom_apply_tmul, ←monomial_one_eq_X_pow],
+    rw [finsupp.sum_single_index]; simp, }
+end
+
+/--
+(Implementation detail)
+
+The equivalence, ignoring the algebra structure, `(A ⊗[R] polynomial R) ≃ polynomial A`.
+-/
+def equiv : (A ⊗[R] polynomial R) ≃ polynomial A :=
+{ to_fun := to_fun_alg_hom R A,
+  inv_fun := inv_fun R A,
+  left_inv := left_inv R A,
+  right_inv := right_inv R A, }
+
+end polynomial_equiv_tensor
+
+open polynomial_equiv_tensor
+
+/--
+The `R`-algebra isomorphism `polynomial A ≃ₐ[R] (A ⊗[R] polynomial R)`.
+-/
+def polynomial_equiv_tensor : polynomial A ≃ₐ[R] (A ⊗[R] polynomial R) :=
+alg_equiv.symm
+{ ..(polynomial_equiv_tensor.to_fun_alg_hom R A), ..(polynomial_equiv_tensor.equiv R A) }
+
+@[simp]
+lemma polynomial_equiv_tensor_apply (p : polynomial A) :
+  polynomial_equiv_tensor R A p =
+    p.eval₂ (include_left : A →ₐ[R] A ⊗[R] polynomial R) ((1 : A) ⊗ₜ[R] (X : polynomial R)) :=
+rfl
+
+@[simp]
+lemma polynomial_equiv_tensor_symm_apply_tmul (a : A) (p : polynomial R) :
+  (polynomial_equiv_tensor R A).symm (a ⊗ₜ p) = p.sum (λ n r, monomial n (a * algebra_map R A r)) :=
+begin
+  simp [polynomial_equiv_tensor, to_fun_alg_hom, alg_hom_of_linear_map_tensor_product, to_fun_linear],
+  refl,
+end
+
+open matrix
+open_locale big_operators
+
+variables {R}
+variables {n : Type w} [fintype n] [decidable_eq n]
+
+/--
+The algebra isomorphism stating "matrices of polynomials are the same as polynomials of matrices".
+
+(You probably shouldn't attempt to use this underlying definition ---
+it's an algebra equivalence, and characterised extensionally by the lemma
+`matrix_polynomial_equiv_polynomial_matrix_coeff_apply` below.)
+-/
+noncomputable def matrix_polynomial_equiv_polynomial_matrix :
+  matrix n n (polynomial R) ≃ₐ[R] polynomial (matrix n n R) :=
```

Comment:
This name is probably getting excessively long, especially as there are lemmas to come which tack more onto the end. Suggestions welcome.

Reply:
Would it be shocking to start abbreviating matrix to mat and/or polynomial to pol or poly? I don't think it would be ambiguous in a linear algebra library. And we already have nhds all over the topology library.

Reply:
I think I will shorten this just to `mat_poly_equiv`, on the basis that what it is equivalent to is fairly obvious?